### PR TITLE
change the main of image_build and write unit test for that

### PR DESCRIPTION
**Pull request by mgold1234**
Signed-off-by: mgold1234 <mgold@redhat.com>

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [x] Tests update
- [ ] Refactor

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

File: pkg/services/images_build/main.go [hunk 1]
```diff
@@ -6,161 +6,143 @@ import (
...
+			mslog.WithField("message", string(e.Value)).Debug("Received an event")
+
+			if e.Headers != nil {
+				mslog.WithField("headers", e.Headers).Debug("Headers received with the event")
+			}
 
-		run := true
-		for run {
-			select {
-			case sig := <-sigchan:
```
Comment (AaronH88) [inline pkg/services/images_build/main.go:77]: same here, why removed?

File: pkg/services/images_build/main.go [hunk 2]
```diff
@@ -6,161 +6,143 @@ import (
...
-		// TODO: and nail record keys to the topic
-		// TODO: make this main.go a single run engine for all microservices
-		topics := []string{kafkacommon.TopicFleetmgmtImageBuild}
-		err = c.SubscribeTopics(topics, nil)
-		if err != nil {
-			mslog.Error("Subscribing to topics failed")
-			// TODO: handle retries
-			// TODO: handle notifications
+	run := true
+	for run {
```
Comment (loadtheaccumulator) [inline pkg/services/images_build/main.go:59]: We should discuss removing the signal channel on L49-50 and the switch that goes with it. It is used to gracefully exit and shutdown when OpenShift terminates a pod.

**Review by loadtheaccumulator — CHANGES_REQUESTED** [review]

File: pkg/services/images_build/main.go [hunk 3]
```diff
@@ -6,161 +6,143 @@ import (
...
+		mslog.WithField("error", "No kafka configuration found")
+		os.Exit(1)
+	}
 
-	if cfg.KafkaConfig.Brokers != nil {
-		consumerGroup := "imagesbuild"
+	consumerGroup := "imagesbuild"
+	c, err := edgeAPIServices.ConsumerService.GetConsumer(consumerGroup)
 
-		sigchan := make(chan os.Signal, 1)
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply pkg/services/images_build/main.go:49]: I thought it should be the same as in image_iso, there we don't have a signal,
can you explain to me what is the difference, please?

File: pkg/services/images_build/main.go (hunk 1 above)
Comment (mgold1234) [reply pkg/services/images_build/main.go:77]: add it back

File: pkg/services/images_build/main.go (hunk 2 above)
Comment (mgold1234) [reply pkg/services/images_build/main.go:64]: I added it back

**Review by tomershinhar — DISMISSED** [review]
LGTM

**Review by loadtheaccumulator — DISMISSED** [review]

**Review by tomershinhar — APPROVED** [review]
LGTM

**Comment by ldjebran** [comment]
remove do not merge as automation is ok now

**Comment by ldjebran** [comment]
remove do not merge as automation is ok now

**Review by loadtheaccumulator — APPROVED** [review]

**Review by loadtheaccumulator — CHANGES_REQUESTED** [review]
Does poll return *kafka.Error or kafka.Error?

File: pkg/services/images_build/main.go [hunk 4]
```diff
@@ -160,8 +148,14 @@ func main() {
 				}
 			}
 		}
-
-		log.Info("Closing consumer\n")
-		c.Close()
 	}
+	log.Info("Closing consumer")
+}
+func main() {
```
Comment (loadtheaccumulator) [inline pkg/services/images_build/main.go:154]: It needs an empty line to separate the functions.

**Review by loadtheaccumulator — CHANGES_REQUESTED** [review]

**Comment by mgold1234** [comment]
@loadtheaccumulator  poll return *kafka.Error

File: pkg/services/images_build/images_build_suite_test.go [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestClients(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Clients Suite")
```
Comment (ldjebran) [inline pkg/services/images_build/images_build_suite_test.go:12]:
```go
RunSpecs(t, "Images Build Suite")
```

File: pkg/services/images_build/main_test.go [hunk 6]
```diff
@@ -1,9 +1,142 @@
...
+						Commit: &models.Commit{
+							OSTreeCommit: faker.UUIDHyphenated(),
+							OrgID:        common.DefaultOrgID,
+							Status:       "BUILDING",
+						},
+						Status:     models.ImageStatusSuccess,
+						ImageSetID: &imageSet.ID,
+						Version:    1,
+						OrgID:      common.DefaultOrgID,
+						Name:       gomock.Any().String(),
```
Comment (ldjebran) [inline pkg/services/images_build/main_test.go:95]: may be this should be the same name as the image set one

File: pkg/services/images_build/main_test.go [hunk 7]
```diff
@@ -1,9 +1,142 @@
...
+					}
+					kafkaMessage := kafka.Message{
+						Key:            key1,
+						Value:          message,
+						TopicPartition: tp,
+					}
+					mockConsumerService.EXPECT().GetConsumer(consumerGroupID).Return(mockConsumer, nil)
+					mockConsumer.EXPECT().SubscribeTopics(gomock.Any(), gomock.Any())
+					mockConsumer.EXPECT().Poll(timeout).Return(&kafkaMessage)
+					mockConsumer.EXPECT().Commit().AnyTimes()
```
Comment (ldjebran) [inline pkg/services/images_build/main_test.go:134]: why .AnyTimes() should't be one time, or I am missing something ? 
```go
mockConsumer.EXPECT().Commit().Times(1)
```

Comment (ldjebran) [reply pkg/services/images_build/main_test.go:134]: as we have one message

File: pkg/services/images_build/main_test.go [hunk 8]
```diff
@@ -1,9 +1,142 @@
...
+					kafkaError := kafka.NewError(kafka.ErrAllBrokersDown, "Error", true)
+					mockConsumer.EXPECT().Poll(timeout).Return(&kafkaError).AnyTimes()
```
Comment (ldjebran) [inline pkg/services/images_build/main_test.go:136]: same here about AnyTime(), may be I am missing something ?

**Review by ldjebran — COMMENTED** [review]
looks good, left some comments.

**Comment by loadtheaccumulator** [comment]
> @loadtheaccumulator poll return *kafka.Error

@mgold1234 Referencing confluent's .poll test, it appears to return kafka.Error.

https://github.com/confluentinc/confluent-kafka-go/blob/8f59eb87ad41e55c87ca80c40ac984eef7126f42/kafka/consumer_test.go#L491-L526

**Comment by mgold1234** [comment]
@loadtheaccumulator when I run it, It's look like it is *kafka.Error.
and it is the same as we have in image_iso/main.go no?
image

**Comment by mgold1234** [comment]
@loadtheaccumulator should I change it in image_iso as well?
thank you

**Comment by loadtheaccumulator** [comment]
> when I run it, It's look like it is *kafka.Error.

@mgold1234 When you run it, I'm assuming it is following the return defined in the mock or the actual method. If the actual method is looking for non-pointer, and because type pointer to type is different than the type itself, the failure will never get caught by the switch in production and only in the unit test. The lines of code I pasted above are from the confluent test for the same library.

**Review by ldjebran — COMMENTED** [review]
lgtm, except the problem with kafka.Error, I agree with @loadtheaccumulator,
, when you supply &kafka.Error from the mock of course you will will get *kafka.Error, 
you have to return back to kafka code and check the error returned, for me when investigating I found the returned is kafka.Error
The basic one in kaka code 
```go

// NewError creates a new Error.
func NewError(code ErrorCode, str string, fatal bool) (err Error) {
	return Error{code: code, str: str, fatal: fatal}
}
```

**Comment by mgold1234** [comment]
@loadtheaccumulator I already change the code to kafka.error can you please look?

**Review by ldjebran — DISMISSED** [review]
lgtm

File: pkg/services/images_build/main.go [hunk 9]
```diff
@@ -161,8 +148,15 @@ func main() {
 				}
 			}
 		}
-
-		log.Info("Closing consumer\n")
-		c.Close()
 	}
+	log.Info("Closing consumer")
```
Comment (loadtheaccumulator) [inline pkg/services/images_build/main.go:152]: Is the consumer explicitly closed somewhere?

**Review by loadtheaccumulator — CHANGES_REQUESTED** [review]

File: pkg/services/images_build/main.go (hunk 9 above)
Comment (loadtheaccumulator) [reply pkg/services/images_build/main.go]: @mgold1234 You should be able to cover that with a mockConsumer.EXPECT() line for .Close in the test.

**Review by ldjebran — APPROVED** [review]

**Review by loadtheaccumulator — APPROVED** [review]

**Pull request merged**